File: backend/app/matching.py
"""Matching engine — connect a surplus crop to nearby processing units.

Given a crop's surplus (in tonnes for the coming week), find units that can process
that crop, sort by distance, and greedily fill their weekly capacity. For each match
we compute the offer price to the farmer and the impact (rupees saved vs dumping,
kg rescued, units engaged).

Plain-Python haversine, no geo libraries — matches the dependency-light style.
"""
from math import radians, sin, cos, asin, sqrt

# Farmer offer floor: we lift the crashed price up to a fair floor.
OFFER_BONUS_PER_KG = 5.0   # ₹/kg above the crash price
MIN_OFFER = 9.0            # ₹/kg absolute floor


def haversine_km(lat1, lng1, lat2, lng2):
    lat1, lng1, lat2, lng2 = map(radians, (lat1, lng1, lat2, lng2))
    dlat = lat2 - lat1
    dlng = lng2 - lng1
    a = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlng / 2) ** 2
    return round(2 * 6371 * asin(sqrt(a)), 1)


def offer_price(crash_price):
    return round(max(MIN_OFFER, crash_price + OFFER_BONUS_PER_KG), 1)


def match(crop_slug, crop_lat, crop_lng, surplus_tonnes, crash_price, units, max_km=80):
    """units: list of dicts with keys name, kind, lat, lng, crops(list), weekly_capacity,
    products, slug, contact. Returns (matches, totals)."""
    offer = offer_price(crash_price)

    candidates = []
    for u in units:
        if crop_slug not in (u.get("crops") or []):
            continue
        dist = haversine_km(crop_lat, crop_lng, u["lat"], u["lng"])
        if dist > max_km:
            continue
        candidates.append((dist, u))
    candidates.sort(key=lambda t: t[0])

    matches = []
    remaining = surplus_tonnes
    for dist, u in candidates:
        if remaining <= 0:
            break
        take = min(u["weekly_capacity"], remaining)
        if take <= 0:
            continue
        remaining -= take
        kg = take * 1000
        rupees_saved = round((offer - crash_price) * kg)
        matches.append({
            "unit_slug": u["slug"],
            "unit_name": u["name"],
            "kind": u["kind"],
            "products": u.get("products") or [],
            "distance_km": dist,
            "allocated_tonnes": round(take, 1),
            "offer_price": offer,
            "rupees_saved": rupees_saved,
            "lat": u["lat"],
            "lng": u["lng"],
        })

    totals = {
        "surplus_tonnes": round(surplus_tonnes, 1),
        "tonnes_matched": round(surplus_tonnes - max(0, remaining), 1),
        "tonnes_unmatched": round(max(0, remaining), 1),
        "kg_rescued": round((surplus_tonnes - max(0, remaining)) * 1000),
        "rupees_saved": sum(m["rupees_saved"] for m in matches),
        "units_engaged": len(matches),
        "offer_price": offer,
        "crash_price": crash_price,
    }
    return matches, totals

File: frontend/src/lib/engine/matching.ts
// Matching engine (ported from backend/app/matching.py). Connect a surplus crop to the
// nearest processing units by distance + capacity, and compute the rupee/kg impact.

import type { Unit } from "./seed";

const OFFER_BONUS_PER_KG = 5.0;
const MIN_OFFER = 9.0;

export function haversineKm(lat1: number, lng1: number, lat2: number, lng2: number) {
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return Math.round(2 * 6371 * Math.asin(Math.sqrt(a)) * 10) / 10;
}

export function offerPrice(crashPrice: number) {
  return Math.round(Math.max(MIN_OFFER, crashPrice + OFFER_BONUS_PER_KG) * 10) / 10;
}

export type Match = {
  unitSlug: string;
  unitName: string;
  kind: string;
  products: string[];
  distanceKm: number;
  allocatedTonnes: number;
  offerPrice: number;
  rupeesSaved: number;
  lat: number;
  lng: number;
};

export type MatchTotals = {
  surplusTonnes: number;
  tonnesMatched: number;
  tonnesUnmatched: number;
  kgRescued: number;
  rupeesSaved: number;
  unitsEngaged: number;
  offerPrice: number;
  crashPrice: number;
};

export function match(
  cropSlug: string,
  cropLat: number,
  cropLng: number,
  surplusTonnes: number,
  crashPrice: number,
  units: Unit[],
  maxKm = 80
): { matches: Match[]; totals: MatchTotals } {
  const offer = offerPrice(crashPrice);

  const candidates = units
    .filter((u) => (u.crops || []).includes(cropSlug))
    .map((u) => ({ dist: haversineKm(cropLat, cropLng, u.lat, u.lng), u }))
    .filter((c) => c.dist <= maxKm)
    .sort((a, b) => a.dist - b.dist);

  const matches: Match[] = [];
  let remaining = surplusTonnes;
  for (const { dist, u } of candidates) {
    if (remaining <= 0) break;
    const take = Math.min(u.weeklyCapacity, remaining);
    if (take <= 0) continue;
    remaining -= take;
    const kg = take * 1000;
    matches.push({
      unitSlug: u.slug,
      unitName: u.name,
      kind: u.kind,
      products: u.products || [],
      distanceKm: dist,
      allocatedTonnes: Math.round(take * 10) / 10,
      offerPrice: offer,
      rupeesSaved: Math.round((offer - crashPrice) * kg),
      lat: u.lat,
      lng: u.lng,
    });
  }

  const matchedTonnes = surplusTonnes - Math.max(0, remaining);
  const totals: MatchTotals = {
    surplusTonnes: Math.round(surplusTonnes * 10) / 10,
    tonnesMatched: Math.round(matchedTonnes * 10) / 10,
    tonnesUnmatched: Math.round(Math.max(0, remaining) * 10) / 10,
    kgRescued: Math.round(matchedTonnes * 1000),
    rupeesSaved: matches.reduce((a, m) => a + m.rupeesSaved, 0),
    unitsEngaged: matches.length,
    offerPrice: offer,
    crashPrice,
  };
  return { matches, totals };
}
